[DTensor] allow OpStrategy to represent ops whose return type is a tuple

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116439
* #116437
* #116428
* #115689
* #115683
* __->__ #115682

**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

XilunWu committed:
[DTensor] allow None op strategy in TupleStrategy for ops that may return None values

[ghstack-poisoned]

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093620).

wanchaol commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -129,15 +129,16 @@ class TupleStrategy(StrategyType):
     """
 
-    def __init__(self, childs: Sequence[StrategyType]) -> None:
+    def __init__(self, childs: Sequence[Optional[StrategyType]]) -> None:
```
did you try to run lintrunner locally and whether it passes the lint checks? I was expecting we have bunch of callsites we need to add asserts to satisfy mypy

XilunWu replied:
Yes I have run local lintrunner and it reports no issue. 
1) I also expected to add asserts but found there're already existing assertions checking types. So there is no need.
2) I checked the lint error in CI and it's unrelated. But obviously here the length issue was not captured by the CI linter so there could be something going wrong.

wanchaol reviewed (approved):
sounds good, please make sure lint is happy before landing, the current linting is flaky

XilunWu committed:
Update on "[DTensor] allow None op strategy in TupleStrategy for ops that may return None values"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows TupleStrategy to have `None` object in `childs` which correspond to
`None` return value in tuple.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] allow OpStrategy to represent ops whose return type is a tuple"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] allow OpStrategy to represent ops whose return type is a tuple"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -65,7 +69,69 @@ class PlacementStrategy:
+                    output_spec_str.append(str(spec))
+            return "(" + ", ".join(output_spec_str) + ")"
+
+    def pretty_print_output_spec_placements(self) -> str:
```
why we have two functions?

* pretty_print_output_spec
* pretty_print_output_spec_placements?

What's the difference here?

I think we should only expose 1 function like before: `pretty_print_placements`

XilunWu replied:
I want to keep `pretty_print_output_spec` because in `OpSchema.__str__` we need to get the `output_spec` string. Using `pretty_print_placements` to do that is quite verbose (10 extra lines in that `elif` branch).

wanchaol reviewed (commented):
Please see inline comments about the changes we need to do

wanchaol reviewed (changes requested):
requesting changes since it was accepted but the PR re-purposed later

XilunWu commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -216,10 +216,23 @@ def spec_to_strategy(spec: object) -> object:
                 for idx, input_spec in enumerate(op_schema.args_spec):
                     desired_spec = (
-                        output_strategy.output_spec
+                        _get_output_spec_from_op_strategy(output_strategy, idx)
```
So we can just replace `strategy.output_spec` with the newly added `strategy.out_spec` here, right?

XilunWu commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -54,9 +54,13 @@ class PlacementStrategy:
     """
 
-    output_spec: DTensorSpec
+    output_spec: Union[DTensorSpec, Sequence[Optional[DTensorSpec]]]
```
Shall we unify this with `OutputSpecType`? The main difference is `OutputSpecType` can be `None`.

XilunWu committed:
Update on "[DTensor] allow OpStrategy to represent ops whose return type is a tuple"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

XilunWu commented:
@wanchaol I addressed most of the comments except that I kept `pretty_print_output_spec` because in `OpSchema.__str__` we need to get the `output_spec` string and this logic involves a check on spec type which cannot be represented using `pretty_print_placements`. However, I plan to see if I can combine `pretty_print_placements` with `pretty_print_output_spec` into one util function. I'll put that attempt in a follow up PR.

XilunWu committed:
Update on "[DTensor] allow OpStrategy to represent ops whose return type is a tuple"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] allow OpStrategy to represent ops whose return type is a tuple"


**Summary**:
Ops like `native_layer_norm_backward` return a tuple of optional torch.Tensor.
This PR allows to use OpStrategy to represent `native_layer_norm_backward`'s
return value sharding.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin wanchaol fduwjj wz337 tianyu-l wconstab yf225

Differential Revision: [D52093620](https://our.internmc.facebook.com/intern/diff/D52093620)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -54,9 +55,13 @@ class PlacementStrategy:
     """
 
-    output_spec: DTensorSpec
+    output_spec: Union[DTensorSpec, Tuple[Optional[DTensorSpec], ...]]
```
No need to do in this PR, I feel maybe as a follow up PR, we can refactor this to a cleaner way:

1. rename `output_spec` to `output_specs` to indicate that there can be multiple output returns
2. rename `out_spec` cached property to `output_spec` to align with `input_specs`

XilunWu replied:
#116437

wanchaol reviewed (approved):
lgtm!

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093620).

XilunWu commented:
@XilunWu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52093620).

XilunWu commented:
@pytorchbot merge

XilunWu commented:
@pytorchbot merge
